src/components/structures/RoomView.js

```diff
@@ -504,17 +526,21 @@ module.exports = React.createClass({
         }
     },
 
-    _paginateCompleted: function() {
-        debuglog("paginate complete");
-
-        // we might have switched rooms since the paginate started - just bin
+    _onTimelineUpdated: function(gotResults) {
+        // we might have switched rooms since the load started - just bin
         // the results if so.
         if (this.unmounted) return;
 
         this.setState({
-            room: MatrixClientPeg.get().getRoom(this.props.roomId),
             paginating: false,
         });
+
+        if (gotResults) {
+            this.setState({
+                events: this._timelineWindow.getEvents(),
```

Comment:
a few reasons:

* `this._timelineWindow.getEvents()` is not a simple getter: it goes and re-builds the event array based on what's in the window right now. Storing that array somewhere is useful, as it makes things like the read-receipt handling much simpler than having to do the double-nested loop that `TimelineWindow.getEvents()` does.
* It feels to me that this is the react-like way of doing things - we *want* to have to call `this.setState` to trigger the re-render when events arrive, rather than having to call `forceUpdate` to tell react there have been changes to state it can't see. Maybe I'm misunderstanding, though.
* This is the only place we need to call `setState({ events: whatever })` in the whole component. I reject your claim it adds more book-keeping ;)